analytics/src/main/java/com/segment/analytics/Analytics.java

```diff
@@ -1394,12 +1398,21 @@ ProjectSettings getSettings() {
     return downloadedSettings;
   }
 
-  void performInitializeIntegrations(ProjectSettings projectSettings) {
+  void performInitializeIntegrations(ProjectSettings projectSettings) throws IOException {
+    if (isNullOrEmpty(projectSettings)) {
+      throw new IOException("ProjectSettings is empty!");
+    }
     ValueMap integrationSettings = projectSettings.integrations();
+    if (isNullOrEmpty(integrationSettings)) {
+      throw new IOException("There are no integration settings!");
+    }
     integrations = new LinkedHashMap<>(factories.size());
     for (int i = 0; i < factories.size(); i++) {
       Integration.Factory factory = factories.get(i);
       String key = factory.key();
+      if (isNullOrEmpty(key)) {
```

Comment:
Instead of a runtime check, we should annotate the Factory method to explicitly note that this should return a NonNull value (example https://github.com/segmentio/analytics-android/blob/93c7d5bb09b593440a2347a6089db3e9f74012da/analytics/src/main/java/com/segment/analytics/integrations/GroupPayload.java#L46). This would move the check to compile time instead of runtime.

Reply:
that appears to be marked @nonnull already.

Reply:
@bsneed yes, that was an example. Specifically here, we should mark the `Integration.factory.key` method as NonNull (https://github.com/segmentio/analytics-android/blob/master/analytics/src/main/java/com/segment/analytics/integrations/Integration.java#L49)

Reply:
Ah, I misunderstood